### PR TITLE
Print ParamTy span when accessing a field (#52082)

### PR DESCRIPTION
**Pull request by Baranowski**

### CODE REVIEW

**Comment by Baranowski** [comment]
I'm not sure the wording of the note printed is correct since I'm not yet familiar with the rustc nomenclature. I might have also missed some other conventions.

File: src/librustc_typeck/check/expr.rs [hunk 1]
```diff
@@ -1491,6 +1502,23 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         err.emit();
     }
 
+    fn explain_param(
```
Comment (Centril) [inline src/librustc_typeck/check/expr.rs]: formatting nit: the stuff until `{` probably fits on a single line.

File: src/librustc_typeck/check/expr.rs [hunk 2]
```diff
@@ -1405,6 +1402,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 _ => {}
             }
 
+            let deref_t = match expr_t.kind {
+                ty::Ref(_, ref_t, _) => ref_t,
+                _ => &expr_t
+            };
+            match deref_t.kind {
```
Comment (Centril) [inline src/librustc_typeck/check/expr.rs]:
```suggestion
            match expr_t.peel_refs().kind {
```
(under the assumption that we want to peel off as many references as possible)

File: src/librustc_typeck/check/expr.rs [hunk 3]
```diff
@@ -1405,6 +1402,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                ty::Adt(def, _) if !def.is_enum() => {
+                    self.suggest_fields_on_recordish(&mut err, def, field);
+                }
+                ty::Param(param_ty) => {
+                    self.explain_param(&mut err, param_ty);
+                }
+                _ => {}
+            }
+
             if field.name == kw::Await {
```
Comment (Centril) [inline src/librustc_typeck/check/expr.rs]: Btw, I added this bit of code and I think it should perhaps be made into its own method as well... if you want to, that could be some nice drive-by cleanup. :)

File: src/librustc_typeck/check/expr.rs [hunk 4]
```diff
@@ -1491,6 +1502,23 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+        let generics = self.tcx.generics_of(self.body_id.owner_def_id());
+        let param_def_id = generics.type_param(&param, self.tcx).def_id;
+        let param_hir_id = match self.tcx.hir().as_local_hir_id(param_def_id) {
+            Some(x) => x,
+            None    => return,
+        };
+        let param_span = self.tcx.hir().span(param_hir_id);
+        let param_name = self.tcx.hir().ty_param_name(param_hir_id);
+
+        err.span_note(param_span, &format!("Type parameter '{}' was declared here", param_name));
```
Comment (Centril) [inline src/librustc_typeck/check/expr.rs]:
```suggestion
        err.span_note(param_span, &format!("type parameter `{}` was declared here", param_name));
```

Comment (estebank) [inline src/librustc_typeck/check/expr.rs]:
```suggestion
        err.span_label(param_span, &format!("type parameter '{}' declared here", param_name));
```

File: src/librustc_typeck/check/expr.rs (hunk 1 above)
Comment (estebank) [inline src/librustc_typeck/check/expr.rs]:
```suggestion
    fn point_at_param_definition(
```

**Review by estebank — APPROVED** [review]
LGTM. Can you make the requested changes? After that and with a green CI, r=me.

File: src/test/ui/typeck/issue-52082.rs [hunk 5]
```diff
@@ -0,0 +1,54 @@
+// Fix issue 52082: Confusing error if accidentially defining a type paramter with the same name as
```
Comment (Centril) [inline src/test/ui/typeck/issue-52082.rs]: Would be helpful to give the filename a bit more context, `issue-52082-type-param-shadows-existing-type.rs`

File: src/librustc_typeck/check/expr.rs (hunk 2 above)
Comment (Baranowski) [reply src/librustc_typeck/check/expr.rs]: Peeling off all the references is consistent with `expr_t.is_primitive_ty()` which got us into this branch (a few lines above), so I suspect we do. Will change.

**Comment by Baranowski** [comment]
There is one problem with the current code. The label in error message:

```
error[E0609]: no field `await` on type `impl Future<Output = ()>`
  --> /home/glon/Programs/rust/src/test/ui/async-await/suggest-switching-edition-on-await.rs:40:7
   |
LL | fn await_on_apit(x: impl Future<Output = ()>) {
   |                     ------------------------ type parameter 'impl Future<Output = ()>' declared here
LL |     x.await;
   |       ^^^^^
```

seems confusing and incorrect. This is issued for the following function in `test/ui/async-await/suggest-switching-edition-on-await.rs`:

```
fn await_on_apit(x: impl Future<Output = ()>) {                                                     
    x.await;                                                                                        
    //~^ ERROR no field `await` on type                                                             
    //~| NOTE to `.await` a `Future`, switch to Rust 2018                                           
    //~| HELP set `edition = "2018"` in `Cargo.toml`                                                
    //~| NOTE for more on editions, read https://doc.rust-lang.org/edition-guide                    
}
```

I will try to dig into `GenericParamDef` etc. and see how to filter out such situations, but if you have some suggestions off the top of your head, I would appreciate them.
